packages/provider/src/utils.ts

```diff
@@ -112,10 +113,123 @@ function recoverSigner(digest: Arrayish, sig: ArcadeumDecodedSigner) {
   }
 }
 
+export async function isValidSignature(
+  address: string,
+  digest: Uint8Array,
+  sig: string,
+  provider?: Provider,
+  arcadeumContext?: ArcadeumContext,
+  chainId?: number
+) {
+  return (
+    isValidEIP721Signature(address, digest, sig) ||
+    isValidEthSignSignature(address, digest, sig) ||
+    (await isValidWalletSignature(address, digest, sig, provider)) ||
+    (await isValidArcadeumWalletSignature(address, digest, sig, provider)) ||
+    // Arcadeum fixed signatures must be validated after wallet signatures
+    // in case the Arcadeum wallet has been updated and has new owners
+    (await isValidArcadeumFixedSignature(address, digest, sig, arcadeumContext, provider, chainId))
+  )
+}
+
+export function isValidEIP721Signature(address: string, digest: Uint8Array, sig: string) {
+  try {
+    return compareAddr(
+      ethers.utils.recoverAddress(
+        digest,
+        ethers.utils.splitSignature(sig)
+      ),
+      address
+    ) === 0
+  } catch {
+    return false
+  }
+}
+
+export function isValidEthSignSignature(address: string, digest: Uint8Array, sig: string) {
+  try {
+    const subDigest = ethers.utils.keccak256(
+      ethers.utils.solidityPack(
+        ['string', 'bytes32'],
+        ['\x19Ethereum Signed Message:\n32', digest]
+      )
+    )
+    return compareAddr(
+      ethers.utils.recoverAddress(
+        subDigest,
+        ethers.utils.splitSignature(sig)
+      ),
+      address
+    ) === 0
+  } catch {
+    return false
+  }
+}
+
+export async function isValidWalletSignature(
+  address: string,
+  digest: Uint8Array,
+  sig: string,
+  provider?: Provider
+) {
+  if (!provider) throw new Error('Wallet signatures require RPC Provider')
+  try {
+    if ((await provider.getCode(address)) === '0x') {
+      return false
+    }
+
+    const wallet = new ethers.Contract(address, erc1271Abi, provider)
+    const response = await wallet.isValidSignature(digest, sig)
+    return erc1271returns.isValidSignatureBytes32 === response
+  } catch {
+    return false
+  }
+}
+
+export async function isValidArcadeumWalletSignature(
+  address: string,
+  digest: Uint8Array,
+  sig: string,
+  provider?: Provider
+) {
+  if (!provider) throw new Error('Arcadeum wallet signatures require RPC Provider')
+  try {
+    const chainId = (await provider.getNetwork()).chainId
```

Comment:
What if provider is on mainnet but we also want to verify on, say, POA or another network?

Reply:
Yep! Good catch